### PR TITLE
Remove virtualenv environment if we encounter unexpected error

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #8306

## What changes are proposed in this pull request?

Remove a virtualenv environment if we encounter an unexpected error while creating it.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/pyfunc/test_virtualenv.py [hunk 1]
```diff
@@ -134,6 +134,27 @@ def test_differenet_requirements_create_different_environments(temp_mlflow_env_r
     assert len(list(temp_mlflow_env_root.iterdir())) == 2
 
 
+@use_temp_mlflow_env_root
+def test_environment_directory_is_cleaned_up_when_unexpected_error_occurs(
+    temp_mlflow_env_root, sklearn_model
+):
+    sklearn_req = "scikit-learn==999.999.999"
```
Comment (BenWilson2) [inline tests/pyfunc/test_virtualenv.py:141]: 83 years from now someone is going to wonder why this test fails ;)

File: mlflow/utils/file_utils.py [hunk 2]
```diff
@@ -810,3 +811,26 @@ def read_chunk(path: os.PathLike, size: int, start_byte: int = 0) -> bytes:
         if start_byte > 0:
             f.seek(start_byte)
         return f.read(size)
+
+
+@contextmanager
+def remove_on_error(path: os.PathLike, onerror=None):
```
Comment (BenWilson2) [inline mlflow/utils/file_utils.py:817]: This is really clever

File: tests/pyfunc/test_virtualenv.py (hunk 1 above)
Comment (harupy) [reply tests/pyfunc/test_virtualenv.py:141]: I don't think that will happen. You normally don't release 999 minor/micro versions :)

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**